core/shared/src/main/scala/fs2/internal/Scope.scala

```diff
@@ -0,0 +1,295 @@
+            case Some((effect, ec)) =>
+              val ref = new async.Ref[F,Unit]()(effect, ec)
+
+              F.flatMap(state.modify { _.copy(refReleased = Some(ref.setAsyncPure(()))) }) { c =>
+                if (c.now.refCount == 0) finishClose(asyncSupport) // refCount were updated before we registered ref signal.
+                else F.flatMap(ref.get) { _ => finishClose(asyncSupport) }
+              }
+          }
+        }
+      }
+    }
+  }
+
+  /**
+    * Close current scope.
+    *
+    * This will acquire all child scopes and resources from the child scopes. Once acquired, this will run
+    * all finalizers for the scopes.
+    *
+    * If scope or any of its child scopes are in process of acquiring asynchronous resources, this will wait till
+    * the resource allocation is completed or will cancel the resource allocation, and then will run the cleanups for that resources.
+    *
+    * Returns either failure or next scope that was not yet closed (ancestor), that may be used as current scope.
+    *
+    * @param asyncSupport If defined, this allows asynchronous execution of cleanup. Always None for pure `F` processes.
+    * @return
+    */
+  private[internal] def close(asyncSupport: Option[(Effect[F],ExecutionContext)]): F[Either[Throwable,Unit]] = {
+    def runAll(sofar: Option[Throwable], finalizers: Catenable[F[Unit]]): F[Either[Throwable,Unit]] = finalizers.uncons match {
+      case None => F.pure(sofar.toLeft(()))
+      case Some((h, t)) => F.flatMap(F.attempt(h)) { res => runAll(sofar orElse res.fold(Some(_), _ => None), t) }
+    }
+    F.flatMap(closeAndReturnFinalizers(asyncSupport)) { finalizers =>
+      runAll(None, finalizers.map(_._2))
+    }
+  }
+
+
+  /**
+    * Returns closes open parent scope or root.
+    */
+  private[internal] def openAncestor: F[Scope[F]] = {
+    self.parent.fold(F.pure(self)) { parent =>
+    F.flatMap(parent.state.get) { s =>
+      if (s.open_?) F.pure(parent)
+      else parent.openAncestor
+    }}
+  }
+
+
+  /**
+    * Allows to acquire `this` scope as resource to `remote` stream.
+    *
+    * Acquisition of the scope prevents this scope to be fully closed (and that means to release its resources) until
+    * returned release of the scope is evaluated.
+    *
+    * This is useful in scenarios, where two streams depends on each other (like Stream#join) and required resources
+    * should not to be released unless all dependent streams will release.
+    *
+    * Linking is implemented as incrementing the `refCount` in `this` scope by one. The `remote` side will
+    * then register supplied callback as their normal resource provided returned callback as release of that resource.
+    *
+    * When this yields to None, that indicates that linking was not succesfull, due `this` scope being closed.
+    *
+    * Note that acquisition of the scope is synchronous, so it is safe to await its evaluation.
+    *
+    */
+  def acquire: F[Option[F[Unit]]] = {
```

Comment:
Can we name this something else? I think the name is too close to the resource management functions above (e.g. beginAcquire)

Reply:
how about acquireScope?